azurerm/internal/services/storage/storage_account_resource.go

```diff
@@ -1004,6 +1110,48 @@ func resourceStorageAccountUpdate(d *schema.ResourceData, meta interface{}) erro
 		}
 	}
 
+	if d.HasChange("routing_preference") {
+		opts := storage.AccountUpdateParameters{
+			AccountPropertiesUpdateParameters: &storage.AccountPropertiesUpdateParameters{
+				RoutingPreference: expandArmStorageAccountRoutingPreference(d.Get("routing_preference").([]interface{})),
+			},
+		}
+
+		if _, err := client.Update(ctx, resourceGroupName, storageAccountName, opts); err != nil {
+			return fmt.Errorf("updating Azure Storage Account routing_preference %q: %+v", storageAccountName, err)
+		}
+	}
+
+	// azure_files_identity_based_authentication must be the last to be updated, cause it'll occupy the storage account for several minutes after receiving the response 200 OK. Issue: https://github.com/Azure/azure-rest-api-specs/issues/11272
+	if d.HasChange("azure_files_identity_based_authentication") {
+		// due to service issue: https://github.com/Azure/azure-rest-api-specs/issues/12473, we need to update to None before changing its DirectoryServiceOptions
+		dsNone := storage.AccountUpdateParameters{
+			AccountPropertiesUpdateParameters: &storage.AccountPropertiesUpdateParameters{
+				AzureFilesIdentityBasedAuthentication: &storage.AzureFilesIdentityBasedAuthentication{
+					DirectoryServiceOptions: storage.DirectoryServiceOptionsNone,
+				},
+			},
+		}
+
+		if _, err := client.Update(ctx, resourceGroupName, storageAccountName, dsNone); err != nil {
+			return fmt.Errorf("updating Azure Storage Account azure_files_identity_based_authentication %q: %+v", storageAccountName, err)
+		}
```

Comment:
To avoid disruptively removing the configuration (or unnecessarily setting to `None` twice) if `DirectoryServiceOptions` has not changed:
```suggestion
		old, new := d.GetChange("azure_files_identity_based_authentication.0.directory_service_options")
		if old != new && new != string(storage.DirectoryServiceOptionsNone) {
			log.Print("[DEBUG] Disabling AzureFilesIdentityBasedAuthentication prior to changing DirectoryServiceOptions")

			dsNone := storage.AccountUpdateParameters{
				AccountPropertiesUpdateParameters: &storage.AccountPropertiesUpdateParameters{
					AzureFilesIdentityBasedAuthentication: &storage.AzureFilesIdentityBasedAuthentication{
						DirectoryServiceOptions: storage.DirectoryServiceOptionsNone,
					},
				},
			}

			if _, err := client.Update(ctx, resourceGroupName, storageAccountName, dsNone); err != nil {
				return fmt.Errorf("updating Azure Storage Account azure_files_identity_based_authentication %q: %+v", storageAccountName, err)
			}
		}
```